### PR TITLE
Update Hello and Swap to authenticated calls

### PR DESCRIPTION
**Pull request by fadeev**
## Call

```bash
npx hardhat compile --force
npx hardhat deploy --gateway 0x6c533f7fe93fae114d0954697069df33c9b74fd7 --network zeta_testnet --name Universal
npx hardhat deploy --gateway 0x0c487a766110c85d301d96e33579c5b317fa4995 --network base_sepolia --name Connected
```

```
🔑 Using account: 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32

🚀 Successfully deployed "Universal" contract on zeta_testnet.
📜 Contract address: 0x496CD66950a1F1c5B02513809A2d37fFB942be1B

🔑 Using account: 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32

🚀 Successfully deployed "Connected" contract on base_sepolia.
📜 Contract address: 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed
```

## Call from Base to ZetaChain

```
npx hardhat connected-call \
  --contract 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed \
  --receiver 0x496CD66950a1F1c5B02513809A2d37fFB942be1B \
  --network base_sepolia \
  --tx-options-gas-price 20000 --types '["string"]' alice
```

https://sepolia.basescan.org/tx/0x133cdf3a06195de0a6bb89dd4761ca98d1301534b3c4987f0ff93c95c3fff78c

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x133cdf3a06195de0a6bb89dd4761ca98d1301534b3c4987f0ff93c95c3fff78c

## Call from ZetaChain to Base

```
npx hardhat universal-call \
  --contract 0x496CD66950a1F1c5B02513809A2d37fFB942be1B \
  --receiver 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed \
  --zrc20 0x236b0DE675cC8F46AE186897fCCeFe3370C9eDeD \
  --function "hello(string)" \
  --network zeta_testnet \
  --tx-options-gas-price 200000 --types '["string"]' alice
```

https://zetachain-testnet.blockscout.com/tx/0x19d476fa2c3d29ba41467ae7f2742541fd11e0b67d6548fe7655a3d40274323e

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x19d476fa2c3d29ba41467ae7f2742541fd11e0b67d6548fe7655a3d40274323e

## Withdraw And Call from ZetaChain to Base

```
npx hardhat universal-withdraw-and-call \
  --contract 0x496CD66950a1F1c5B02513809A2d37fFB942be1B \
  --receiver 0x775329c70A8d09AEb5e5ca92C369FF3155C5f1Ed \
  --zrc20 0x236b0DE675cC8F46AE186897fCCeFe3370C9eDeD \
  --function "hello(string)" \
  --amount 0.005 \
  --network zeta_testnet \
  --types '["string"]' hello
```

https://zetachain-testnet.blockscout.com/tx/0x67099389ab6cb44ee03602d164320c615720b57762c5ddab042d65bdbe30c7a2

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x67099389ab6cb44ee03602d164320c615720b57762c5ddab042d65bdbe30c7a2

## Swap

```
npx hardhat compile --force 

npx hardhat deploy --name Swap --gateway 0x6c533f7fe93fae114d0954697069df33c9b74fd7 --system-contract 0xEdf1c3275d13489aCdC6cD6eD246E72458B8795B --network zeta_testnet

🚀 Successfully deployed contract on zeta_testnet.
📜 Contract address: 0xf4fE3154c6dBb5492D3A4BbcF6A68878DF61C8A0
```

```
npx hardhat evm-deposit-and-call \
  --receiver 0xf4fE3154c6dBb5492D3A4BbcF6A68878DF61C8A0 \
  --amount 0.01 \
  --network bsc_testnet \
  --gateway-evm 0x0c487a766110c85d301d96e33579c5b317fa4995 \
  --types '["address", "bytes"]' 0x05BA149A7bd6dC1F937fA9046A9e05C05f3b18b0 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32
```

https://testnet.bscscan.com/tx/0x0d944bbfd5c9f3a9ffba14d342c30808a0e61e9cd6e7889ea6e557a266f3527d

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x0d944bbfd5c9f3a9ffba14d342c30808a0e61e9cd6e7889ea6e557a266f3527d

https://sepolia.basescan.org/tx/0x990d2f4c9dcbb96f0b61f6a9d06d6680e45cbdc6f4af9ab4fff2ba0207c9a199

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x990d2f4c9dcbb96f0b61f6a9d06d6680e45cbdc6f4af9ab4fff2ba0207c9a199

https://amoy.polygonscan.com/tx/0xb162f4a9a2414a7fd63ef3e7d34a113af14d325a97cab1cc8036d7d958eafae9

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0xb162f4a9a2414a7fd63ef3e7d34a113af14d325a97cab1cc8036d7d958eafae9

## NFT

```
npx hardhat compile --force

npx hardhat deploy --gateway 0x6c533f7fe93fae114d0954697069df33c9b74fd7 --network zeta_testnet --gas-limit 500000 --name Universal

🚀 Successfully deployed "Universal" contract on zeta_testnet.
📜 Contract address: 0xCD58653a854817692079efd8ef0f6928aA66F628

npx hardhat deploy --gateway 0x0c487a766110c85d301d96e33579c5b317fa4995 --network bsc_testnet --name Connected

🚀 Successfully deployed "Connected" contract on bsc_testnet.
📜 Contract address: 0x31F76a088dF9802A283088cC8976f2395944712A

npx hardhat deploy --gateway 0x0c487a766110c85d301d96e33579c5b317fa4995 --network base_sepolia --name Connected

🚀 Successfully deployed "Connected" contract on base_sepolia.
📜 Contract address: 0xe36eA381567Ac954D3d068EC7033724B545E5Ae2

npx hardhat connected-set-counterparty --network bsc_testnet --contract 0x31F76a088dF9802A283088cC8976f2395944712A --counterparty 0xCD58653a854817692079efd8ef0f6928aA66F628 --json

{"contractAddress":"0x31F76a088dF9802A283088cC8976f2395944712A","universalContract":"0xCD58653a854817692079efd8ef0f6928aA66F628","transactionHash":"0x7a7ba679db5210ce6b8290e20883749e0601408162b43bc37a10b6ee7146919b"}

npx hardhat connected-set-counterparty --network base_sepolia --contract 0xe36eA381567Ac954D3d068EC7033724B545E5Ae2 --counterparty 0xCD58653a854817692079efd8ef0f6928aA66F628 --json

{"contractAddress":"0xe36eA381567Ac954D3d068EC7033724B545E5Ae2","universalContract":"0xCD58653a854817692079efd8ef0f6928aA66F628","transactionHash":"0xe9b4039e138f7c068ca288a403a6a7499454286323b4b9b6cc12c93e80e88561"}

npx hardhat universal-set-counterparty --network zeta_testnet --contract 0xCD58653a854817692079efd8ef0f6928aA66F628 --counterparty 0x31F76a088dF9802A283088cC8976f2395944712A --zrc20 0xd97B1de3619ed2c6BEb3860147E30cA8A7dC9891 --json

{"contractAddress":"0xCD58653a854817692079efd8ef0f6928aA66F628","zrc20":"0xd97B1de3619ed2c6BEb3860147E30cA8A7dC9891","connectedContractAddress":"0x31F76a088dF9802A283088cC8976f2395944712A","transactionHash":"0x27e7c5a1878eca4fce74405e1fe722c7ef3660abfb5bc51c9b8912ceb65b5a5b"}

npx hardhat universal-set-counterparty --network zeta_testnet --contract 0xCD58653a854817692079efd8ef0f6928aA66F628 --counterparty 0xe36eA381567Ac954D3d068EC7033724B545E5Ae2 --zrc20 0x236b0DE675cC8F46AE186897fCCeFe3370C9eDeD --json

{"contractAddress":"0xCD58653a854817692079efd8ef0f6928aA66F628","zrc20":"0x236b0DE675cC8F46AE186897fCCeFe3370C9eDeD","connectedContractAddress":"0xe36eA381567Ac954D3d068EC7033724B545E5Ae2","transactionHash":"0x9ffc05de29db59a719bde8bc3f6de43baf660af1e5702e21f367ebfb5916c25c"}

npx hardhat mint --network zeta_testnet --json --contract 0xCD58653a854817692079efd8ef0f6928aA66F628 --token-uri https://example.com/nft/metadata/1

{"contractAddress":"0xCD58653a854817692079efd8ef0f6928aA66F628","mintTransactionHash":"0x76f359263dfcccd898e7564f102d54b11daa763028cc725ee5bd3f05bcc6c2c9","recipient":"0x4955a3F38ff86ae92A914445099caa8eA2B9bA32","tokenURI":"https://example.com/nft/metadata/1","tokenId":"456003645905194763481523353718654221867725429194"}

```

### ZetaChain -> Base NFT transfer

```
npx hardhat transfer --network zeta_testnet --json --token-id 456003645905194763481523353718654221867725429194 --from 0xCD58653a854817692079efd8ef0f6928aA66F628 --to 0x236b0DE675cC8F46AE186897fCCeFe3370C9eDeD
```

```json
{"contractAddress":"0xCD58653a854817692079efd8ef0f6928aA66F628","transferTransactionHash":"0x554dec1282f3c367b4a465e391a0887f5b891f86ff2f32048c16c4932c278c4c","sender":"0x4955a3F38ff86ae92A914445099caa8eA2B9bA32","tokenId":"456003645905194763481523353718654221867725429194"}
```

https://zetachain-testnet.blockscout.com/tx/0x554dec1282f3c367b4a465e391a0887f5b891f86ff2f32048c16c4932c278c4c

https://zetachain-athens.blockpi.network/lcd/v1/public/zeta-chain/crosschain/inboundHashToCctxData/0x554dec1282f3c367b4a465e391a0887f5b891f86ff2f32048c16c4932c278c4c

## Release Notes

- **New Features**
	- Introduced new tasks for deploying and interacting with contracts, including `connected-deposit`, `connected-deposit-and-call`, and `universal-withdraw`.
	- Added a new `localnet.sh` script for managing local blockchain interactions.
	- Enhanced the `Connected` and `Universal` contracts with improved error handling and gas limit management.
	- Updated README files with new tutorial links for better user guidance.
	- Implemented new `Swap` and `SwapToAnyToken` contracts for facilitating cross-chain token swaps.
	- Updated deployment scripts to include new parameters for Uniswap integration.
	- Added support for managing gas limits in cross-chain operations.
	- Updated the `localnet.sh` scripts across various examples to improve deployment logic and parameter handling.
	- Renamed tasks and parameters to reflect updated terminology, including changes from "counterparty" to "universal" in relevant tasks.
	- Removed obsolete files and scripts related to the omnichain swap functionality.

- **Bug Fixes**
	- Improved error handling in various tasks to ensure better feedback during contract interactions.

- **Chores**
	- Updated dependency versions across multiple `package.json` files to ensure compatibility and access to the latest features.
	- Added `.gitignore` entries for `localnet.json` across multiple directories to manage ignored files.

### CODE REVIEW

**Comment by fadeev** [comment]
For some reason Slither cannot find outdated config and CI fails.

**Comment by fadeev** [comment]
@andresaiello please, review.

File: examples/call/contracts/Connected.sol [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    GatewayEVM public immutable gateway;
+
+    event RevertEvent(string, RevertContext);
+    event HelloEvent(string, string);
+
+    error TransferFailed();
+    error ApprovalFailed();
+
+    modifier onlyGateway() {
+        require(msg.sender == address(gateway), "Caller is not the gateway");
```
Comment (andresaiello) [inline examples/call/contracts/Connected.sol:17]: transfer and approval are with custom errors and this with require. move all to the same (I prefer custom errors)

File: examples/call/contracts/Connected.sol [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        gateway.deposit{value: msg.value}(receiver, revertOptions);
+    }
+
+    function deposit(
+        address receiver,
+        uint256 amount,
+        address asset,
+        RevertOptions memory revertOptions
+    ) external {
+        if (!IERC20(asset).transferFrom(msg.sender, address(this), amount)) {
```
Comment (andresaiello) [inline examples/call/contracts/Connected.sol:46]: maybe we should use safeTransferFrom

File: examples/call/contracts/Connected.sol [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    function depositAndCall(
+        address receiver,
+        uint256 amount,
+        address asset,
+        bytes calldata message,
+        RevertOptions memory revertOptions
+    ) external {
+        if (!IERC20(asset).transferFrom(msg.sender, address(this), amount)) {
```
Comment (andresaiello) [inline examples/call/contracts/Connected.sol:62]: same, let's use safeTransferFrom

File: examples/call/contracts/Universal.sol [hunk 4]
```diff
@@ -22,46 +22,56 @@ contract Hello is UniversalContract {
...
+        address zrc20,
+        RevertOptions memory revertOptions
+    ) external {
+        (address gasZRC20, uint256 gasFee) = IZRC20(zrc20).withdrawGasFee();
+        uint256 target = zrc20 == gasZRC20 ? amount + gasFee : amount;
+        if (!IZRC20(zrc20).transferFrom(msg.sender, address(this), target))
+            revert TransferFailed();
+        IZRC20(zrc20).approve(address(gateway), target);
+        if (zrc20 != gasZRC20) {
+            if (
```
Comment (andresaiello) [inline examples/call/contracts/Universal.sol:53]: in the previus transfer you use if() { revert } and in this one if() revert. Let's use the same in all. I prefer with {} to make it more readeable

File: examples/call/contracts/Connected.sol (hunk 1 above)
Comment (fadeev) [reply examples/call/contracts/Connected.sol]: c89c6da

File: examples/call/contracts/Connected.sol (hunk 2 above)
Comment (fadeev) [reply examples/call/contracts/Connected.sol]: a6d9809

File: examples/call/contracts/Connected.sol (hunk 3 above)
Comment (fadeev) [reply examples/call/contracts/Connected.sol]: a6d9809

File: examples/call/contracts/Universal.sol (hunk 4 above)
Comment (fadeev) [reply examples/call/contracts/Universal.sol:57]: Do you use curly braces even for one-liner `if`s?

Comment (fadeev) [reply examples/call/contracts/Universal.sol:57]: A:

```solidity
if (gas == 0) revert InvalidGasLimit();
```

or B:

```solidity
if (gas == 0) revert InvalidGasLimit();
```

Comment (fadeev) [reply examples/call/contracts/Universal.sol:57]: 83ab10b

**Pull request merged**